src/ModulusRemainder.cpp

```diff
@@ -306,11 +330,24 @@ ModulusRemainder unify_alternatives(ModulusRemainder a, ModulusRemainder b) {
         << "unified modulus   = " << modulus << "\n"
         << "unified remainder = " << ra << "\n";
 
-    return ModulusRemainder(modulus, ra);
+    return {modulus, ra};
+}
+
+ModulusRemainder ModulusRemainder::intersect(const ModulusRemainder &a, const ModulusRemainder &b) {
+    // We have x == ma * y + ra == mb * z + rb
+
+    // For coprime ma and mb you want to use the Chinese remainder
+    // theorem. In our case, the moduli will almost always be
```

Comment:
Shouldn't we check that though, at least as an assert, if not just bailing to ModulusRemainder{}?

Reply:
I should make the comment clearer. The code is conservative. It's supposed to produce a set which contains at least the intersection of the two sets A and B. Returning the smaller of A and B satisfies this, and is sufficiently tight for the cases we care about.

Reply:
I must be misunderstanding this function. What about ma = 2, ra = 0 and mb = 2, rb = 1?

Reply:
The intersection is the empty set. A contains the empty set, so returning A satisfies the requirement of the function.

This would show up if you have a load that you know is aligned as 2x + 1, and then later you discover evidence that it's aligned as 2x + 0, so you take the intersection... i.e. the only way you'd find yourself in that situation is if there's a bug somewhere else.

Reply:
That said, there's no reason not to make the function produce a tighter bounding of the true intersection.

Reply:
> A contains the empty set, so returning A satisfies the requirement of the function.

This is the part I must be missing. So this function returns a set that contains the intersection, but also contains other stuff that isn't the intersection?

It's fine if that's the intended behavior, but the comment in the header really needs to explain this better. This is the opposite of "conservative" to me :) To me, "conservative" would be a set that contains only the intersection of the two sets, but is missing some things that are also in the intersection.

Reply:
Or to say that the way I should have said more succinctly: I would have thought a conservative intersection was a subset of the intersection, not a superset.

Reply:
Clarified comments.